<a href="https://colab.research.google.com/github/MohammedRafik0909/Data-Science-Projects/blob/main/Target_And_Actual.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PREDICTIVE ANALYSIS USING THE TARGET VS ACTUAL SALES AND QUANTITY **

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [3]:
from google.colab import files
uploaded =  files.upload()

Saving TargetAndSales_upd.csv to TargetAndSales_upd (1).csv


IMPORTING DATA

In [4]:
data = pd.read_csv('TargetAndSales_upd.csv')

SEEING THE FIRST FIVE ROWS FROM THE DATASET

In [5]:
data.head()

Date         Floor        Division           Section  \
0  01/02/2024  SECOND FLOOR           Girls  Girls Inner Wear   
1  01/02/2024  SECOND FLOOR       Foot Wear   Girls Foot Wear   
2  01/02/2024   FIRST FLOOR       Foot Wear      Luggage Bags   
3  01/02/2024   THIRD FLOOR            Boys  Boys Accessories   
4  01/02/2024   FIRST FLOOR  Ladies Textile   Designer Sarees   

              Department         Supervisor  Merchandiser  Target Value  \
0              Giw-Brief      MR.DANDHAPANI  MR.MANIMARAN        1779.4   
1          Gfw-Pu Sandal  MR.SAYAD MOHAMMED  MR.ASHIK ALI         690.1   
2         Lgb-Travel Bag  MR.SAYAD MOHAMMED  MR.ASHIK ALI        6725.0   
3          Bac-Boys Mask  MR.SAYAD MOHAMMED  MR.ASHIK ALI           0.0   
4  Ltds-Designer Dhavani   MR.JAHIR HUSSAIN    MR.SHARIFF           0.0   

   Target Qty  Actual Value  Actual Qty  
0        31.9         898.0        17.0  
1         2.7         695.0         3.0  
2         5.9        3175.5         3.0  
3         0.0           0.0         0.0  
4         0.0           0.0         0.0

# DATA PREPROCESSING

CHECKING FOR MISSING VALUE AND IT'S APPEARS TO BE ZERO MISSING VALUE

In [6]:
data.isnull().sum()

Date            0
Floor           0
Division        0
Section         0
Department      0
Supervisor      0
Merchandiser    0
Target Value    0
Target Qty      0
Actual Value    0
Actual Qty      0
dtype: int64

In [7]:
data.fillna(method ='ffill',inplace = True)

In [8]:
data.isnull().sum()

Date            0
Floor           0
Division        0
Section         0
Department      0
Supervisor      0
Merchandiser    0
Target Value    0
Target Qty      0
Actual Value    0
Actual Qty      0
dtype: int64

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49391 entries, 0 to 49390
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          49391 non-null  object 
 1   Floor         49391 non-null  object 
 2   Division      49391 non-null  object 
 3   Section       49391 non-null  object 
 4   Department    49391 non-null  object 
 5   Supervisor    49391 non-null  object 
 6   Merchandiser  49391 non-null  object 
 7   Target Value  49391 non-null  float64
 8   Target Qty    49391 non-null  float64
 9   Actual Value  49391 non-null  float64
 10  Actual Qty    49391 non-null  float64
dtypes: float64(4), object(7)
memory usage: 4.1+ MB


In [10]:
data.describe()

Target Value    Target Qty   Actual Value    Actual Qty
count   49391.000000  49391.000000   49391.000000  49391.000000
mean     6641.012051     18.557381    3398.553178      8.633188
std     15134.622422     39.371204    8978.906953     20.039538
min         0.000000      0.000000   -9370.000000    -28.000000
25%        28.600000      0.100000       0.000000      0.000000
50%       811.600000      2.600000      47.700000      1.000000
75%      5735.200000     17.600000    2550.000000      7.000000
max    214119.900000    509.000000  169467.400000    437.000000

In [11]:
from sklearn.preprocessing import LabelEncoder

CONVERTING OBJECT DATA TYPE TO DATE TIME DATATYPE

In [12]:
data['Date'] = pd.to_datetime(data['Date'], dayfirst=True)

In [13]:
print(data['Date'])

0       2024-02-01
1       2024-02-01
2       2024-02-01
3       2024-02-01
4       2024-02-01
           ...    
49386   2024-05-31
49387   2024-05-31
49388   2024-05-31
49389   2024-05-31
49390   2024-05-31
Name: Date, Length: 49391, dtype: datetime64[ns]


EXTRACTING YEAR,MONTH,DAY AND WEEK FROM THE DATE FOR TIMESTAMP ANALYSIS

In [14]:
data['Year'] = data['Date'].dt.year
data['Month'] = data['Date'].dt.month
data['Day'] = data['Date'].dt.day
data['DayOfWeek'] = data['Date'].dt.dayofweek

In [15]:
data = data.drop(columns=['Date'])

In [16]:
data.head(10)

Floor              Division           Section  \
0    SECOND FLOOR                 Girls  Girls Inner Wear   
1    SECOND FLOOR             Foot Wear   Girls Foot Wear   
2     FIRST FLOOR             Foot Wear      Luggage Bags   
3     THIRD FLOOR                  Boys  Boys Accessories   
4     FIRST FLOOR        Ladies Textile   Designer Sarees   
5    SECOND FLOOR                Infant   Infant Separate   
6    GROUND FLOOR        Ladies Textile        Silk Saree   
7  BASEMENT FLOOR  Furniture & Fittings          Clothing   
8    SECOND FLOOR                 Girls  Girls Loung Wear   
9    SECOND FLOOR             Foot Wear    Kids Foot Wear   

              Department         Supervisor    Merchandiser  Target Value  \
0              Giw-Brief      MR.DANDHAPANI    MR.MANIMARAN        1779.4   
1          Gfw-Pu Sandal  MR.SAYAD MOHAMMED    MR.ASHIK ALI         690.1   
2         Lgb-Travel Bag  MR.SAYAD MOHAMMED    MR.ASHIK ALI        6725.0   
3          Bac-Boys Mask  MR.SAYAD MOHAMMED    MR.ASHIK ALI           0.0   
4  Ltds-Designer Dhavani   MR.JAHIR HUSSAIN      MR.SHARIFF           0.0   
5          Isp-Cot Frock     MR.SURULIRAJAN    MR.MANIMARAN         940.4   
6     Ltsk-Soft Pure Mix     MR.TAMILARASAN  MR.TAMILARASAN        5017.5   
7          Ffbc-Velvette      MR.MANIKANDAN   MR.PRABAKARAN           0.0   
8        Glw-Lounge Wear      MR.DANDHAPANI    MR.MANIMARAN           0.0   
9         Kfw-Kids Shoes  MR.SAYAD MOHAMMED    MR.ASHIK ALI         794.7   

   Target Qty  Actual Value  Actual Qty  Year  Month  Day  DayOfWeek  
0        31.9         898.0        17.0  2024      2    1          3  
1         2.7         695.0         3.0  2024      2    1          3  
2         5.9        3175.5         3.0  2024      2    1          3  
3         0.0           0.0         0.0  2024      2    1          3  
4         0.0           0.0         0.0  2024      2    1          3  
5         2.4           0.0         0.0  2024      2    1          3  
6         1.3           0.0         0.0  2024      2    1          3  
7         0.0           0.0         0.0  2024      2    1          3  
8         0.0           0.0         0.0  2024      2    1          3  
9         3.5        1055.0         5.0  2024      2    1          3

LABEL ENCODERS AND FIT TRANSFORM OF DATA FOR CHANGING CATERGORICAL DATA TO NUMERICAL.

In [17]:
label_encoders = {}
for column in data.select_dtypes(include=['object']).columns:
    label_encoders[column] = LabelEncoder()
    data[column] = label_encoders[column].fit_transform(data[column])

In [18]:
data.head(10)

Floor  Division  Section  Department  Supervisor  Merchandiser  \
0      5         4       27          87           2             6   
1      5         2       25          78          11             3   
2      2         2       43         170          11             3   
3      6         0        3           1          11             3   
4      2         7       17         259           4            10   
5      5         5       34         127          14             6   
6      4         7       55         277          15            13   
7      0         3       15          60           6             8   
8      5         4       28          91           2             6   
9      5         2       35         139          11             3   

   Target Value  Target Qty  Actual Value  Actual Qty  Year  Month  Day  \
0        1779.4        31.9         898.0        17.0  2024      2    1   
1         690.1         2.7         695.0         3.0  2024      2    1   
2        6725.0         5.9        3175.5         3.0  2024      2    1   
3           0.0         0.0           0.0         0.0  2024      2    1   
4           0.0         0.0           0.0         0.0  2024      2    1   
5         940.4         2.4           0.0         0.0  2024      2    1   
6        5017.5         1.3           0.0         0.0  2024      2    1   
7           0.0         0.0           0.0         0.0  2024      2    1   
8           0.0         0.0           0.0         0.0  2024      2    1   
9         794.7         3.5        1055.0         5.0  2024      2    1   

   DayOfWeek  
0          3  
1          3  
2          3  
3          3  
4          3  
5          3  
6          3  
7          3  
8          3  
9          3

# FEATURE ENGINEERING

In [19]:
X = data.drop(columns=['Actual Value', 'Actual Qty'])#FEATURE VARIABLE
y_value = data['Actual Value']#TARGET VARIABLE
y_qty = data['Actual Qty']

In [20]:
print(X)

       Floor  Division  Section  Department  Supervisor  Merchandiser  \
0          5         4       27          87           2             6   
1          5         2       25          78          11             3   
2          2         2       43         170          11             3   
3          6         0        3           1          11             3   
4          2         7       17         259           4            10   
...      ...       ...      ...         ...         ...           ...   
49386      3         9       65         363          11             3   
49387      3         9       65         364          11             3   
49388      3         9       65         365          11             3   
49389      3         9       65         366          11             3   
49390      3         9       61         357           5             0   

       Target Value  Target Qty  Year  Month  Day  DayOfWeek  
0            1779.4        31.9  2024      2    1          3

In [21]:
print("Actual Value and Actual Qty")
print(y_value,y_qty)

Actual Value and Actual Qty
0         898.0
1         695.0
2        3175.5
3           0.0
4           0.0
          ...  
49386       0.0
49387       0.0
49388       0.0
49389       0.0
49390       0.0
Name: Actual Value, Length: 49391, dtype: float64 0        17.0
1         3.0
2         3.0
3         0.0
4         0.0
         ... 
49386     0.0
49387     0.0
49388     0.0
49389     0.0
49390     0.0
Name: Actual Qty, Length: 49391, dtype: float64


# DATA SPLITTING

In [22]:
X_train, X_test, y_value_train, y_value_test, y_qty_train, y_qty_test = train_test_split(X, y_value, y_qty, test_size=0.2, random_state=42)

In [23]:
print(X_train)

       Floor  Division  Section  Department  Supervisor  Merchandiser  \
24854      5         6       38         190          11             3   
41241      2         6       36         140          10             4   
16380      4        10       47         384           6             8   
33175      5         2       66         413          11             3   
48550      3         9       48         337           5             0   
...      ...       ...      ...         ...         ...           ...   
11284      5         2        5           8          11             3   
44732      2         6       41         198          10             4   
38158      5         2       43         159          11             3   
860        6         0        8          18           7             1   
15795      2         7       17         259           4            10   

       Target Value  Target Qty  Year  Month  Day  DayOfWeek  
24854          16.0         0.3  2024      4    2          1

# **DATA MODELLING**

RandomForestRegressor for predicting Actual Sales
AND XGBRegressor for predicting Actual Qty

In [24]:
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

In [25]:
model_value = XGBRegressor(n_estimators=100, learning_rate=0.1, random_state=42)
model_qty = RandomForestRegressor(n_estimators=100, random_state=42)

In [26]:
model_value.fit(X_train, y_value_train)
model_qty.fit(X_train, y_qty_train)

RandomForestRegressor(random_state=42)

In [27]:
y_value_pred = model_value.predict(X_test)
y_qty_pred = model_qty.predict(X_test)

In [28]:
print(y_value_pred)

[22660.92      107.67341   122.04235 ... 15336.373    -170.08878
   908.1517 ]


EVALLUVATION WITH REGRESSION METRICS LIKE MAE,MSE,RMSE AND R2

In [36]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

def evaluate_model(y_true, y_pred, model_name):
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = mean_squared_error(y_true, y_pred, squared=False)
    r2 = r2_score(y_true, y_pred)

    print(f"Evaluation metrics for {model_name}:")
    print(f"Mean Absolute Error (MAE): {mae}")
    print(f"Mean Squared Error (MSE): {mse}")
    print(f"Root Mean Squared Error (RMSE): {rmse}")
    print(f"R-squared (R2): {r2}")

evaluate_model(y_value_test, y_value_pred, 'Actual Value')
evaluate_model(y_qty_test, y_qty_pred, 'Actual Quantity')


Evaluation metrics for Actual Value:
Mean Absolute Error (MAE): 1720.4297376618533
Mean Squared Error (MSE): 18867441.76561715
Root Mean Squared Error (RMSE): 4343.666857117055
R-squared (R2): 0.7592842879128722
Evaluation metrics for Actual Quantity:
Mean Absolute Error (MAE): 3.4656837736613024
Mean Squared Error (MSE): 63.828497218544385
Root Mean Squared Error (RMSE): 7.9892738855633425
R-squared (R2): 0.8329645437766013


THE PREDICTED MODEL ACHIEVED 76% FOR ACTUAL SALES AND 83% FOR ACTUAL QTY

In [42]:
input_instance = {
    'Floor': 'FIRST FLOOR',
    'Division': 'Girls',
    'Section': 'Girls Inner Wear',
    'Department': 'Giw-Brief',
    'Supervisor': 'MR.DANDHAPANI',
    'Merchandiser': 'MR.MANIMARAN',
    'Target Value': 1779.4,
    'Target Qty': 31.9,
    'Year': 2024,
    'Month': 7,
    'Day': 1,
    'DayOfWeek': 3
}


#input_instance.pop('Actual Value', None)
#input_instance.pop('Actual Qty', None)

for column in ['Floor', 'Division', 'Section', 'Department', 'Supervisor', 'Merchandiser']:
    input_instance[column] = label_encoders[column].transform([input_instance[column]])[0]


input_data = pd.DataFrame(input_instance, index=[0])


predicted_value = model_value.predict(input_data)
predicted_qty = model_qty.predict(input_data)


print("Predicted Actual Value:", predicted_value[0])
print("Predicted Actual Qty:", predicted_qty[0])


Predicted Actual Value: 1937.0063
Predicted Actual Qty: 24.947


EVALUVATING WITH SAMPLE INSTANCES

EG:'Floor': 'SECOND FLOOR',
    'Division': 'Girls',
    'Section': 'Girls Inner Wear',
    'Department': 'Giw-Brief',
    'Supervisor': 'MR.DANDHAPANI',
    'Merchandiser': 'MR.MANIMARAN',
    'Target Value': 1779.4,
    'Target Qty': 31.9,
    'Year': 2024,
    'Month': 7,
    'Day': 1,
    'DayOfWeek': 3

THE SAMPLE INSTANCES GAVE A PREDICTED SALES OF **1937.0063**
AND FOR PREDICTED QUANTITY OF **24.947**

---



In [50]:
input_instance = {
    'Floor': 'FIRST FLOOR',
    'Division': 'Girls',
    'Section': 'Girls Inner Wear',
    'Department': 'Giw-Brief',
    'Supervisor': 'MR.MANIKANDAN',
    'Merchandiser': 'MR.PRABAKARAN',
    'Target Value': 10000.4,
    'Target Qty': 50.9,
    'Year': 2024,
    'Month': 7,
    'Day': 1,
    'DayOfWeek': 3
}


#input_instance.pop('Actual Value', None)
#input_instance.pop('Actual Qty', None)

for column in ['Floor', 'Division', 'Section', 'Department', 'Supervisor', 'Merchandiser']:
    input_instance[column] = label_encoders[column].transform([input_instance[column]])[0]


input_data = pd.DataFrame(input_instance, index=[0])


predicted_value = model_value.predict(input_data)
predicted_qty = model_qty.predict(input_data)


print("Predicted Actual Value:", predicted_value[0])
print("Predicted Actual Qty:", predicted_qty[0])


Predicted Actual Value: 6759.515
Predicted Actual Qty: 23.175
